File: Core/Entities/OrderAggregate/Order.cs
using System.ComponentModel.DataAnnotations.Schema;

namespace Core.Entities.OrderAggregate;

public class Order : BaseEntity
{
    public string AppUserId { get; set; }

    [NotMapped] private IList<OrderItem> _orderItems;

    public IList<OrderItem> OrderItems
    {
        get => _orderItems;
        set
        {
            _orderItems = value;
            CalculateSubtotal();
            CalculateTax();
            CalculateTotal();
        }
    }

    public ShippingAddress ShippingAddress { get; set; }

    public OrderStatus OrderStatus { get; set; } = OrderStatus.Pending;


    [NotMapped] private DeliveryMethod _deliveryMethod { get; set; }

    public DeliveryMethod DeliveryMethod
    {
        get => _deliveryMethod;
        set
        {
            _deliveryMethod = value;
            CalculateTotal();
        }
    }

    public int DeliveryMethodId { get; set; }


    public DateTime OrderDate { get; set; } = DateTime.UtcNow;

    [NotMapped] public decimal Subtotal { get; set; }
    [NotMapped] public decimal Tax { get; set; }
    [NotMapped] public decimal Total { get; set; }

    public Order()
    {
    }

    public Order(string appUserId, IList<OrderItem> orderItems, ShippingAddress shippingAddress, DeliveryMethod deliveryMethod)
    {
        AppUserId = appUserId;
        OrderItems = orderItems;
        ShippingAddress = shippingAddress;
        DeliveryMethod = deliveryMethod;
    }

    private void CalculateSubtotal()
    {
        Subtotal = OrderItems.Sum(item => item.Subtotal);
    }

    private void CalculateTax()
    {
        Tax = Subtotal * IConstants.TaxRate;
    }

    private void CalculateTotal()
    {
        Total = Subtotal + Tax + (DeliveryMethod?.Price ?? 0);
    }
}

File: Client/src/app/shared/models/order.ts
import {Address} from "./user";

export class OrderToCreate {
  simpleBasketId: string
  addressDto: Address
  deliveryMethodId: number

  constructor(simpleBasketId: string, addressDto: Address, deliveryMethodId: number) {
    this.simpleBasketId = simpleBasketId;
    this.addressDto = addressDto;
    this.deliveryMethodId = deliveryMethodId;
  }
}

export interface DeliveryMethod {
  id: number
  name: string
  description: string
  approximatedDeliveryDays: string
  price: number
}

export interface Order {
  id: number
  orderItems: OrderItem[]
  shippingAddress: Address
  orderStatus: string
  deliveryMethod: string
  deliveryFee: number
  orderDate: string
  subtotal: number
  tax: number
  total: number
}

export interface OrderItem {
  productId: number
  productName: string
  description: string
  price: number
  pictureUrl: string
  brand: string
  type: string
  quantity: number
  subtotal: number
}

